lib/validator/task/mod.rs

```diff
@@ -768,101 +772,387 @@ pub(in crate::validator) fn disconnect_block(
     let missing_blocks: Vec<BlockHash> = tokio::task::block_in_place(|| {
         let rotxn = dbs.read_txn()?;
-        dbs.block_hashes
+        let ancestor_headers = dbs
+            .block_hashes
             .ancestor_headers(&rotxn, main_tip)
-            .map(|(block_hash, _header)| Ok(block_hash))
-            .take_while(|block_hash| Ok(!dbs.block_hashes.contains_block(&rotxn, block_hash)?))
-            .collect()
-            .map_err(error::Sync::from)
+            .collect::<Vec<_>>()?;
+
+        tracing::info!(
+            main_tip = ?main_tip,
+            "Found {} ancestor headers",
+            ancestor_headers.len()
+        );
+
+        Ok::<_, error::Sync>(
+            ancestor_headers
+                .iter()
+                .map(|(block_hash, _header)| *block_hash)
+                .take_while(|block_hash| {
+                    !dbs.block_hashes
+                        .contains_block(&rotxn, block_hash)
+                        .unwrap_or(false)
+                })
+                .collect::<Vec<_>>(),
+        )
     })?;
     if missing_blocks.is_empty() {
+        tracing::info!("No missing blocks, skipping sync");
         return Ok(());
     }
+    tracing::info!("Syncing {} missing blocks", missing_blocks.len(),);
+    let mut total_blocks_fetched = 0;
     for missing_block in missing_blocks.into_iter().rev() {
-        tracing::debug!("Syncing block `{missing_block}` -> `{main_tip}`");
-        let block = main_client
+        // If we're processing the very first missing block, that means we need to
+        // update the stored chain tip. Otherwise we'll get an error from within
+        // connect_block, as that expects to always work on the current tip.
+        //
+        // This is needed in the case of reorgs. We can end up in a situation
+        // where the stored enforcer tip is /ahead/ of missing_block, and this
+        // needs to be fixed before we can continue.
+        let is_first_block = total_blocks_fetched == 0;
+
+        let block = main_rpc_client
             .get_block(missing_block, U8Witness::<0>)
             .map_err(|err| error::Sync::JsonRpc {
                 method: "getblock".to_owned(),
                 source: err,
             })
             .await?
             .0;
+        total_blocks_fetched += 1;
+
         let mut rwtxn = dbs.write_txn()?;
+
+        let is_genesis = block.header.prev_blockhash == BlockHash::all_zeros();
+        if is_first_block && !is_genesis {
+            let new_chain_tip = block.header.prev_blockhash;
+            let current_tip = dbs.current_chain_tip.try_get(&rwtxn, &UnitKey)?;
+
+            if current_tip != Some(new_chain_tip) {
+                tracing::info!(
+                    "Syncing first block, updating enforcer tip from {} to {}",
+                    current_tip.unwrap_or(BlockHash::all_zeros()),
+                    new_chain_tip,
+                );
+                dbs.current_chain_tip
```

Comment:
This is updated by `connect_block`, should not be done here

Reply:
I added this because it was a point needed in order to sync correctly after a reorg had taken place. When I test now I'm able to do without it... Have to admit this is a bit of a clusterfuck... Removing!